core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java

```diff
@@ -0,0 +1,610 @@
+            } catch (IOException e) {
+                throw new ParsingException(p.getTokenLocation(), "Could not parse inner query definition", e);
+            }
+        }, SearchSourceBuilder.INNER_HITS_FIELD);
+    }
+
+    public static InnerHitBuilder optionalReadFromStream(StreamInput in) throws IOException {
+        if (in.readBoolean()) {
+            return new InnerHitBuilder(in);
+        } else {
+            return null;
+        }
+    }
+
+    private String name;
+    private String nestedPath;
+    private String parentChildType;
+
+    private int from;
+    private int size = 3;
+    private boolean explain;
+    private boolean version;
+    private boolean trackScores;
+
+    private List<String> fieldNames;
+    private QueryBuilder<?> query = new MatchAllQueryBuilder();
+    private List<SortBuilder<?>> sorts;
+    private List<String> fieldDataFields;
+    private List<ScriptField> scriptFields;
+    private HighlightBuilder highlightBuilder;
+    private InnerHitsBuilder innerHitsBuilder;
+    private FetchSourceContext fetchSourceContext;
+
+    // pkg protected, because is used in InnerHitsBuilder
+    InnerHitBuilder(StreamInput in) throws IOException {
+        name = in.readOptionalString();
+        nestedPath = in.readOptionalString();
+        parentChildType = in.readOptionalString();
+        from = in.readVInt();
+        size = in.readVInt();
+        explain = in.readBoolean();
+        version = in.readBoolean();
+        trackScores = in.readBoolean();
+        fieldNames = (List<String>) in.readGenericValue();
+        fieldDataFields = (List<String>) in.readGenericValue();
+        if (in.readBoolean()) {
+            scriptFields = in.readList(t -> ScriptField.PROTOTYPE.readFrom(in));
+        }
+        fetchSourceContext = FetchSourceContext.optionalReadFromStream(in);
+        if (in.readBoolean()) {
+            int size = in.readVInt();
+            sorts = new ArrayList<>(size);
+            for (int i = 0; i < size; i++) {
+                sorts.add(in.readSortBuilder());
+            }
+        }
+        highlightBuilder = in.readOptionalWriteable(HighlightBuilder.PROTOTYPE::readFrom);
+        if (in.readBoolean()) {
+            query = in.readQuery();
+        }
+        innerHitsBuilder = in.readOptionalWriteable(InnerHitsBuilder.PROTO::readFrom);
+    }
+
+    public InnerHitBuilder() {
+    }
+
+    public InnerHitBuilder setParentChildType(String parentChildType) {
+        this.parentChildType = parentChildType;
+        return this;
+    }
+
+    public InnerHitBuilder setNestedPath(String nestedPath) {
+        this.nestedPath = nestedPath;
+        return this;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public InnerHitBuilder setName(String name) {
+        this.name = name;
+        return this;
+    }
+
+    public int getFrom() {
+        return from;
+    }
+
+    public InnerHitBuilder setFrom(int from) {
```

Comment:
I wonder if we should add validation check to these setter methods like we do in the query builders? So for this one check that the value is >= 0 and for setName() make sure the name passed in is not null.